src/event.rs

```diff
@@ -0,0 +1,566 @@
+				let payment_hash: PaymentHash = Readable::read(reader)?;
+				Ok(Self::PaymentSuccessful { payment_hash })
+			}
+			1u8 => {
+				let payment_hash: PaymentHash = Readable::read(reader)?;
+				Ok(Self::PaymentFailed { payment_hash })
+			}
+			2u8 => {
+				let payment_hash: PaymentHash = Readable::read(reader)?;
+				let amount_msat: u64 = Readable::read(reader)?;
+				Ok(Self::PaymentReceived { payment_hash, amount_msat })
+			}
+			3u8 => {
+				let channel_id: [u8; 32] = Readable::read(reader)?;
+				let user_channel_id: u64 = Readable::read(reader)?;
+				Ok(Self::ChannelReady { channel_id, user_channel_id })
+			}
+			4u8 => {
+				let channel_id: [u8; 32] = Readable::read(reader)?;
+				let user_channel_id: u64 = Readable::read(reader)?;
+				Ok(Self::ChannelClosed { channel_id, user_channel_id })
+			}
+			//5u8 => {
+			// TODO OnChainPaymentSent
+			//}
+			//6u8 => {
+			// TODO OnChainPaymentReceived
+			//}
+			_ => Err(lightning::ln::msgs::DecodeError::InvalidValue),
+		}
+	}
+}
+
+impl Writeable for LdkLiteEvent {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), lightning::io::Error> {
+		match self {
+			Self::PaymentSuccessful { payment_hash } => {
+				0u8.write(writer)?;
+				payment_hash.write(writer)?;
+				Ok(())
+			}
+			Self::PaymentFailed { payment_hash } => {
+				1u8.write(writer)?;
+				payment_hash.write(writer)?;
+				Ok(())
+			}
+			Self::PaymentReceived { payment_hash, amount_msat } => {
+				2u8.write(writer)?;
+				payment_hash.write(writer)?;
+				amount_msat.write(writer)?;
+				Ok(())
+			}
+			Self::ChannelReady { channel_id, user_channel_id } => {
+				3u8.write(writer)?;
+				channel_id.write(writer)?;
+				user_channel_id.write(writer)?;
+				Ok(())
+			}
+			Self::ChannelClosed { channel_id, user_channel_id } => {
+				4u8.write(writer)?;
+				channel_id.write(writer)?;
+				user_channel_id.write(writer)?;
+				Ok(())
+			} //Self::OnChainPaymentSent { .. } => {
+			  //TODO
+			  //}
+			  //Self::OnChainPaymentReceived { .. } => {
+			  //TODO
+			  //}
+		}
+	}
+}
+
+pub(crate) struct LdkLiteEventQueue<K: KVStorePersister> {
+	queue: Mutex<EventQueueSerWrapper>,
+	notifier: Condvar,
+	persister: Arc<K>,
+}
+
+impl<K: KVStorePersister> LdkLiteEventQueue<K> {
+	pub(crate) fn new(persister: Arc<K>) -> Self {
+		let queue: Mutex<EventQueueSerWrapper> = Mutex::new(EventQueueSerWrapper(VecDeque::new()));
+		let notifier = Condvar::new();
+		Self { queue, notifier, persister }
+	}
+	pub(crate) fn add_event(&self, event: LdkLiteEvent) -> Result<(), Error> {
+		{
+			let mut locked_queue = self.queue.lock().unwrap();
+			locked_queue.0.push_back(event);
+			self.persister.persist(EVENTS_PERSISTENCE_KEY, &*locked_queue)?;
+		}
+
+		self.notifier.notify_one();
+		Ok(())
+	}
+
+	pub(crate) fn next_event(&self) -> LdkLiteEvent {
+		let locked_queue = self
+			.notifier
+			.wait_while(self.queue.lock().unwrap(), |queue| queue.0.is_empty())
+			.unwrap();
+		locked_queue.0.front().unwrap().clone()
```

Comment:
Should we store / return an `Arc` to avoid cloning?

Reply:
Now reverted this with https://github.com/lightningdevkit/ldk-lite/pull/10/commits/3b6ded0e6c6849621a82c151ccac0db9ab17f31f for two reasons a) I find using a plain `Event` preferable in the interface but even more b) turns out that UniFFI (see #25) doesn't support `Arc`'ed enum return values for the time being. However, it will itself wrap the `Event` in an `Arc` for the exposed interface, but sadly can't re-use the `Arc` in the case of an enum.